avocado/utils/asset.py

```diff
@@ -71,21 +71,178 @@ def __init__(self, name, asset_hash, algorithm, locations, cache_dirs,
 
+        # Break early. We need to know the source location of the asset.
+        if not self.locations:
+            raise EnvironmentError("An asset should have at least one source"
+                                   " location. Double check your asset name"
+                                   " and location and make sure to use a"
+                                   " supported protocol.")
+
+    def _create_hash_file(self, asset_path):
+        """
+        Compute the hash of the asset file and add it to the CHECKSUM
+        file.
+
+        :param asset_file: full path of the asset file.
+        """
+        result = crypto.hash_file(asset_path, algorithm=self.algorithm)
+        with open(self._get_hash_file(asset_path), 'w') as hash_file:
+            hash_file.write('%s %s\n' % (self.algorithm, result))
+
+    def _create_metadata_file(self, asset_dir):
+        """
+        Creates JSON file with metadata.
+        The file will be saved as `asset_file`_metadata.json
+
+        :param asset_dir: full path where the asset file is located.
+        """
+        metadata_file_name = "%s_metadata.json" % self.basename
+        metadata_path = os.path.join(os.path.dirname(asset_dir),
+                                     metadata_file_name)
+        metadata = json.dumps(self.metadata)
+        with open(metadata_path, "w") as metadata_file:
+            metadata_file.write(metadata)
+
+    def _download(self, url_obj, asset_path):
+        """
+        Download the asset from an url.
+
+        :param url_obj: object from urlparse.
+        :param asset_path: full path of the asset file.
+        :returns: if the downloaded file matches the hash.
+        :rtype: bool
+        """
+        try:
+            # Temporary unique name to use while downloading
+            temp = '%s.%s' % (asset_path,
+                              next(tempfile._get_candidate_names()))  # pylint: disable=W0212
+            url_download(url_obj.geturl(), temp)
+
+            # Acquire lock only after download the file
+            with FileLock(asset_path, 1):
+                shutil.move(temp, asset_path)
+                self._create_hash_file(asset_path)
+                return self._verify(asset_path)
+        except OSError:
+            if os.path.isfile(temp):
+                os.remove(temp)
+            elif os.path.isfile(asset_path):
+                os.remove(asset_path)
+
+    @staticmethod
+    def _get_hash_file(asset_path):
+        """
+        Returns the file name that contains the hash for a given asset file
+
+        :param asset_path: full path of the asset file.
+        :returns: the CHECKSUM path
+        :rtype: str
+        """
+        return '%s-CHECKSUM' % asset_path
+
+    def _get_hash_from_file(self, asset_path):
+        """
+        Read the CHECKSUM file from the asset and return the hash.
+
+        :param asset_path: full path of the asset file.
+        :returns: the hash, if it exists.
```

Comment:
The hash always exists because the file is created regardless. What makes this method to return `None` is a malformed hash string (i.e. does not match `'%s [a-f0-9]{32,128}' % self.algorithm`

Reply:
Indeed!